README
NAME
    "Parse::Man" - parse nroff-formatted manpages

DESCRIPTION
    This abstract subclass of Parser::MGC recognises nroff grammar from a
    file or string value. It invokes methods when various nroff directives
    are encountered. It is intended that this class be used as a base class,
    with methods provided to handle the various directives and formatting
    options. Typically a subclass will store intermediate results in a data
    structure, building it as directed by these method invocations.

TEXT CHUNK FORMATTING METHODS
    The following methods are used to parse formatted text. Each is passed a
    plain string value from the input content.

  $parser->chunk_B( $text )
    Handles text content from ".B" directives and "\fB" inline formatting.

  $parser->chunk_I( $text )
    Handles text content from ".I" directives and "\fI" inline formatting.

  $parser->chunk_R( $text )
    Handles text content from ".R" directives and "\fR" inline formatting.

    These above three methods are also used to handle ".BI", ".IB", ".RB",
    ".BR", ".RI" and ".IR" directives.

  $parser->chunk_SM( $text )
    Handles text content from ".SM" directives.

PARAGRAPH HANDLING METHODS
    The following methods are used to form paragraphs out of formatted text
    chunks. Their return values are ignored.

  $parser->para_TH( $name, $section )
    Handles the ".TH" paragraph which gives the page title and section
    number.

  $parser->para_SH( $title )
    Handles the ".SH" paragraph, which gives a section header.

  $parser->para_SS( $title )
    Handles the ".SS" paragraph, which gives a sub-section header.

  $parser->para_TP( $opts )
    Handles a ".TP" paragraph, which gives a term definition.

  $parser->para_IP( $opts )
    Handles a ".IP" paragraph, which is indented like the definition part of
    a ".TP" paragraph.

  $parser->para_P( $opts )
    Handles the ".P", ".PP" or ".LP" paragraphs, which are all synonyms for
    a plain paragraph content.

AUTHOR
    Paul Evans <leonerd@leonerd.org.uk>



Changes
Revision history for Parse-Man

0.01    First version, released on an unsuspecting world.



lib/Parse/Man.pm
#  You may distribute under the terms of either the GNU General Public License
#  or the Artistic License (the same terms as Perl itself)
#
#  (C) Paul Evans, 2011 -- leonerd@leonerd.org.uk

package Parse::Man;

use strict;
use warnings;

use base qw( Parser::MGC );

our $VERSION = '0.01';

use constant pattern_ws => qr/[ \t]+/;

=head1 NAME

C<Parse::Man> - parse nroff-formatted manpages

=head1 DESCRIPTION

This abstract subclass of L<Parser::MGC> recognises F<nroff> grammar from a
file or string value. It invokes methods when various F<nroff> directives are
encountered. It is intended that this class be used as a base class, with
methods provided to handle the various directives and formatting options.
Typically a subclass will store intermediate results in a data structure,
building it as directed by these method invocations.

=cut

sub parse
{
   my $self = shift;

   $self->_change_para_options(
      mode    => "P",
      filling => 1,
      indent  => 0,
   );
   $self->{para_flushed} = 0;

   $self->sequence_of( \&parse_line );
}

sub token_nl
{
   my $self = shift;
   $self->expect( "\n" );
}

sub token_chunk
{
   my $self = shift;

   return $self->any_of(
      sub { ( $self->expect( qr/"((?:\\.|[^"\\\n]+)*)(?:"|$)/m ) )[1] },
      sub { $self->expect( qr/\S+/ ) },
   );
}

sub parse_chunks
{
   my $self = shift;
   @{ $self->sequence_of( \&token_chunk ) };
}

sub parse_chunks_flat
{
   my $self = shift;
   return join " ", $self->parse_chunks;
}

sub parse_line
{
   my $self = shift;

   $self->commit;

   $self->any_of(
      sub {
         my $directive = ( $self->expect( qr/\.([A-Z]+)/i ) )[1];
         $self->commit;
         $self->${\"parse_directive_$directive"}( $self );
         $self->token_nl;
      },
      sub {
         # comments
         $self->expect( qr/\.\\".*?\n/ );
      },
      sub {
         $self->commit;
         $self->scope_of( undef, sub { $self->parse_plaintext }, "\n" );
      },
   );
}

=head1 TEXT CHUNK FORMATTING METHODS

The following methods are used to parse formatted text. Each is passed a plain
string value from the input content.

=cut

=head2 $parser->chunk_B( $text )

Handles text content from C<.B> directives and C<\fB> inline formatting.

=cut

sub parse_directive_B
{
   my $self = shift;
   $self->_flush_para;
   $self->chunk_B( $self->parse_chunks_flat );
}

=head2 $parser->chunk_I( $text )

Handles text content from C<.I> directives and C<\fI> inline formatting.

=cut

sub parse_directive_I
{
   my $self = shift;
   $self->_flush_para;
   $self->chunk_I( $self->parse_chunks_flat );
}

=head2 $parser->chunk_R( $text )

Handles text content from C<.R> directives and C<\fR> inline formatting.

These above three methods are also used to handle C<.BI>, C<.IB>, C<.RB>,
C<.BR>, C<.RI> and C<.IR> directives.

=cut

sub parse_directive_R
{
   my $self = shift;
   $self->_flush_para;
   $self->chunk_R( $self->parse_chunks_flat );
}

=head2 $parser->chunk_SM( $text )

Handles text content from C<.SM> directives.

=cut

sub parse_directive_SM
{
   my $self = shift;
   $self->_flush_para;
   $self->chunk_SM( $self->parse_chunks_flat );
}

sub _parse_directive_alternate
{
   my $self = shift;
   my ( $first, $second ) = @_;
   $self->_flush_para;
   my $i = 0;
   map $self->${\(++$i % 2 ? $first : $second )}( $_ ), $self->parse_chunks;
}

sub parse_directive_BI
{
   my $self = shift;
   $self->_parse_directive_alternate( "chunk_B", "chunk_I" );
}

sub parse_directive_IB
{
   my $self = shift;
   $self->_parse_directive_alternate( "chunk_I", "chunk_B" );
}

sub parse_directive_RB
{
   my $self = shift;
   $self->_parse_directive_alternate( "chunk_R", "chunk_B" );
}

sub parse_directive_BR
{
   my $self = shift;
   $self->_parse_directive_alternate( "chunk_B", "chunk_R" );
}

sub parse_directive_RI
{
   my $self = shift;
   $self->_parse_directive_alternate( "chunk_R", "chunk_I" );
}

sub parse_directive_IR
{
   my $self = shift;
   $self->_parse_directive_alternate( "chunk_I", "chunk_R" );
}

=head1 PARAGRAPH HANDLING METHODS

The following methods are used to form paragraphs out of formatted text
chunks. Their return values are ignored.

=cut

=head2 $parser->para_TH( $name, $section )

Handles the C<.TH> paragraph which gives the page title and section number.

=cut

sub parse_directive_TH
{
   my $self = shift;
   $self->_change_para( "P" ),
   $self->para_TH( $self->parse_chunks );
}

=head2 $parser->para_SH( $title )

Handles the C<.SH> paragraph, which gives a section header.

=cut

sub parse_directive_SH
{
   my $self = shift;
   $self->_change_para( "P" ),
   $self->para_SH( $self->parse_chunks_flat );
}

=head2 $parser->para_SS( $title )

Handles the C<.SS> paragraph, which gives a sub-section header.

=cut

sub parse_directive_SS
{
   my $self = shift;
   $self->_change_para( "P" ),
   $self->para_SS( $self->parse_chunks_flat );
}

=head2 $parser->para_TP( $opts )

Handles a C<.TP> paragraph, which gives a term definition.

=cut

sub parse_directive_TP
{
   my $self = shift;
   $self->_change_para( "TP" );
}

=head2 $parser->para_IP( $opts )

Handles a C<.IP> paragraph, which is indented like the definition part of a
C<.TP> paragraph.

=cut

sub parse_directive_IP
{
   my $self = shift;
   $self->_change_para( "IP" );
}

=head2 $parser->para_P( $opts )

Handles the C<.P>, C<.PP> or C<.LP> paragraphs, which are all synonyms for a
plain paragraph content.

=cut

sub parse_directive_P
{
   my $self = shift;
   $self->_change_para( "P" );
}

{
   no warnings 'once';
   *parse_directive_PP = *parse_directive_LP = \&parse_directive_P;
}

sub parse_directive_RS
{
   my $self = shift;
   $self->_change_para_options( indent => "4n" );
}

sub parse_directive_RE
{
   my $self = shift;
   $self->_change_para_options( indent => "0" );
}

sub parse_directive_br
{
   my $self = shift;
   $self->entity_br;
}

sub parse_directive_fi
{
   my $self = shift;
   $self->_change_para_options( filling => 1 );
}

sub parse_directive_in
{
   my $self = shift;

   my @ret;
   my $indent = 0;

   $self->maybe( sub {
      $indent = $self->expect( qr/[+-]?\d+[n]?/ );
   } );

   $self->_change_para_options( indent => $indent );
}

sub parse_directive_nf
{
   my $self = shift;
   $self->_change_para_options( filling => 0 );
}

sub parse_directive_sp
{
   my $self = shift;
   $self->entity_sp;
}

sub parse_plaintext
{
   my $self = shift;

   my @format = "R";

   $self->_flush_para;

   $self->sequence_of(
      sub { $self->any_of(
         sub { $self->expect( qr/\\fP/ ); pop @format },
         sub { push @format, ( $self->expect( qr/\\f([BIR])/ ) )[1]; },
         sub { my $else = ( $self->expect( qr/\\(.)/ ) )[1]; $self->${\"chunk_$format[-1]"}( $else ) },
         sub { $self->${\"chunk_$format[-1]"}( $self->substring_before( qr/[\\\n]/ ) ) },
      ); }
   );
}

sub _change_para
{
   my $self = shift;
   my ( $mode ) = @_;
   $self->_change_para_options( mode => $mode );
   $self->{para_flushed} = 0;
}

sub _change_para_options
{
   my $self = shift;
   my %opts = @_;

   if( grep { ($self->{para_options}{$_}//"") ne $opts{$_} } keys %opts ) {
      $self->{para_flushed} = 0;
   }

   $self->{para_options}{$_} = $opts{$_} for keys %opts;
}

sub _flush_para
{
   my $self = shift;
   if( !$self->{para_flushed} ) {
      my $mode = $self->{para_options}{mode};
      $self->${\"para_$mode"}( $self->{para_options} );
      $self->{para_flushed}++;
   }
   else {
      $self->join_para;
   }
}

=head1 AUTHOR

Paul Evans <leonerd@leonerd.org.uk>

=cut

0x55AA;


examples/man-to-html.pl
#!/usr/bin/perl

use strict;
use warnings;

use Template;
use String::Tagged::HTML;
use File::Slurp qw( slurp );
use Getopt::Long;

my $FILE_EXTENSION = "html";
my $LINK_EXTENSION;
my $OUTPUT_DIR;
my $PAGE_TEMPLATE;

GetOptions(
   'file-extension|e=s' => \$FILE_EXTENSION,
   'link-extension=s'   => \$LINK_EXTENSION,
   'output-dir|O=s'     => \$OUTPUT_DIR,
   'template|t=s'       => \$PAGE_TEMPLATE,
) or exit(1);

defined $LINK_EXTENSION or $LINK_EXTENSION = $FILE_EXTENSION;
defined $PAGE_TEMPLATE  or $PAGE_TEMPLATE = \*DATA;

package ManToHTMLParser;
use base qw( Parse::Man::DOM );
sub chunklist_class { return "ManToHTMLParser::Chunklist"; }

package ManToHTMLParser::Chunklist;
use base qw( Parse::Man::DOM::Chunklist );

my %format_to_tag = (
   B  => "b",
   I  => "i",
   SM => "small",
);

sub as_html
{
   my $self = shift;

   my $ret = String::Tagged::HTML->new;

   foreach my $chunk ( $self->chunks ) {
      if( $chunk->is_linebreak ) {
         $ret->append( "\n" );
      }
      elsif( $chunk->is_space ) {
         $ret->append( " " );
      }
      elsif( $chunk->is_break ) {
         $ret->append_tagged( "<br/>", raw => 1 );
      }
      elsif( $chunk->format eq "R" ) {
         $ret->append( $chunk->text );
      }
      else {
         my $tag = $format_to_tag{ $chunk->format };
         $ret->append_tagged( $chunk->text, $tag => 1 );
      }
   }

   my $str = $ret->str;

   # Turn foo(1) text into a link
   while( $str =~ m{([[:alpha:]_][[:alnum:]_]*)\((\d[a-z]*)\)}g ) {
      my $name    = $1;
      my $section = $2;

      my $ofs = $-[0];
      my $len = $+[0] - $-[0];

      my $href = ::make_link( lc $name, $section );

      $ret->apply_tag( $ofs, $len, a => { href => $href } ) if defined $href;
   }

   return $ret->as_html;
}

package main;

my @doms;
my %doms_by_namesection;

sub make_link
{
   my ( $name, $section ) = @_;

   if( exists $doms_by_namesection{"$name.$section"} ) {
      return "$name.$section.$LINK_EXTENSION";
   }
   else {
      print STDERR "Omitting link to $name($section) because no file is being generated\n";
      return undef;
   }
}

my $parser = ManToHTMLParser->new;
my $template = Template->new( { DEBUG => 1 } );

my $pagefile = slurp $PAGE_TEMPLATE;

foreach my $manfile ( @ARGV ) {
   my $dom = $parser->from_file( $manfile );

   push @doms, $dom;
   $doms_by_namesection{ lc $dom->meta("name")->value . "." . $dom->meta("section")->value } = $dom;
}

foreach my $dom ( @doms ) {
   my $output;

   unless( $template->process( \$pagefile, { document => $dom }, \$output ) ) {
      print STDERR "Failed: ", $template->error, "\n";
      next;
   }

   my $outfile = join ".", lc $dom->meta("name")->value, $dom->meta("section")->value, $FILE_EXTENSION;

   $outfile = "$OUTPUT_DIR/$outfile" if defined $OUTPUT_DIR;

   my $outh;
   unless( open $outh, ">", $outfile ) {
      print STDERR "Cannot write $outfile - $!\n";
      next;
   }

   print $outh $output;
}

__DATA__
[% BLOCK heading -%]
    <h[% level %]>[% text | html %]</h[% level %]>
[% END -%]
[% BLOCK plain;
   SET tag = para.filling ? "p" : "pre"; -%]
    <[% tag %]>[% para.body.as_html %]</[% tag %]>
[% END -%]
[% BLOCK term -%]
    <dl>
      <dt>[% para.term.as_html %]</dt>
      <dd>[% para.definition.as_html %]</dd>
    </dl>
[% END -%]
[% BLOCK indent -%]
    <dl>
      <dd>[% para.body.as_html %]</dd>
    </dl>
[% END -%]
<html>
  <head>
    <title>[% document.meta("name").value %]</title>
  </head>
  <body>
    <h1>[% document.meta("name").value %] ([% document.meta("section").value %])</h1>
[% FOREACH para IN document.paras;
     SWITCH para.type;
       CASE "heading";
         INCLUDE heading level=para.level+1 text=para.text;
       CASE "plain";
         INCLUDE plain para=para;
       CASE "term";
         INCLUDE term para=para;
       CASE "indent"; 
         INCLUDE indent para=para;
     END;
   END -%]
  </body>
</html>


lib/Parse/Man/DOM.pm
#  You may distribute under the terms of either the GNU General Public License
#  or the Artistic License (the same terms as Perl itself)
#
#  (C) Paul Evans, 2011 -- leonerd@leonerd.org.uk

package Parse::Man::DOM;

use strict;
use warnings;

use base qw( Parse::Man );

our $VERSION = '0.01';

=head1 NAME

C<Parse::Man::DOM> - parse nroff-formatted manpages and return a DOM tree

=head1 SYNOPSIS

 use Parse::Man::DOM;

 my $parser = Parse::Man::DOM->new;

 my $document = $parser->from_file( "my_manpage.1" );

 print "The manpage name is", $document->meta( "name" ), "\n";

=head1 DESCRIPTION

This subclass of L<Parse::Man> returns an object tree representing the parsed
content of the input file. The returned result will be an object of the
C<Parse::Man::DOM::Document> class, which itself will contain other objects
nested within it.

=cut

sub parse
{
   my $self = shift;

   local $self->{current_document} = $self->_make( document => );

   $self->SUPER::parse( @_ );

   return $self->{current_document};
}

sub para_TH
{
   my $self = shift;
   my ( $name, $section ) = @_;

   $self->{current_document}->add_meta( $self->_make( metadata => name    => $name ) );
   $self->{current_document}->add_meta( $self->_make( metadata => section => $section ) );
}

sub para_SH
{
   my $self = shift;
   my ( $text ) = @_;
   $self->{current_document}->append_para( $self->_make( heading => 1 => $text ) );
}

sub para_SS
{
   my $self = shift;
   my ( $text ) = @_;
   $self->{current_document}->append_para( $self->_make( heading => 2 => $text ) );
}

sub para_P
{
   my $self = shift;
   my ( $opts ) = @_;
   $self->{current_document}->append_para( $self->_make( para_plain => $opts, $self->_make( chunklist => ) ) );
}

sub para_TP
{
   my $self = shift;
   my ( $opts ) = @_;
   $self->{current_document}->append_para( $self->_make( para_term => $opts, $self->_make( chunklist => ), $self->_make( chunklist => ) ) );
}

sub para_IP
{
   my $self = shift;
   my ( $opts ) = @_;
   $self->{current_document}->append_para( $self->_make( para_indent => $opts, $self->_make( chunklist => ) ) );
}

sub chunk_R
{
   my $self = shift;
   my ( $text ) = @_;
   $self->{current_document}->append_chunk( $self->_make( chunk => R => $text ) );
}

sub chunk_I
{
   my $self = shift;
   my ( $text ) = @_;
   $self->{current_document}->append_chunk( $self->_make( chunk => I => $text ) );
}

sub chunk_B
{
   my $self = shift;
   my ( $text ) = @_;
   $self->{current_document}->append_chunk( $self->_make( chunk => B => $text ) );
}

sub chunk_SM
{
   my $self = shift;
   my ( $text ) = @_;
   $self->{current_document}->append_chunk( $self->_make( chunk => SM => $text ) );
}

sub join_para
{
   my $self = shift;
   $self->{current_document}->append_chunk( $self->_make( linebreak => ) );
}

sub entity_br
{
   my $self = shift;
   $self->{current_document}->append_chunk( $self->_make( break => ) );
}

sub entity_sp
{
   my $self = shift;
   $self->{current_document}->append_chunk( $self->_make( space => ) );
}

sub _make
{
   my $self = shift;
   my $type = shift;
   my $code = $self->can( "${type}_class" ) or die "Unable to make a ${type}";
   return $code->()->new( @_ );
}

use constant document_class    => "Parse::Man::DOM::Document";
use constant metadata_class    => "Parse::Man::DOM::Metadata";
use constant heading_class     => "Parse::Man::DOM::Heading";
use constant para_plain_class  => "Parse::Man::DOM::Para::Plain";
use constant para_term_class   => "Parse::Man::DOM::Para::Term";
use constant para_indent_class => "Parse::Man::DOM::Para::Indent";
use constant chunklist_class   => "Parse::Man::DOM::Chunklist";
use constant chunk_class       => "Parse::Man::DOM::Chunk";
use constant space_class       => "Parse::Man::DOM::Space";
use constant break_class       => "Parse::Man::DOM::Break";
use constant linebreak_class   => "Parse::Man::DOM::Linebreak";

package Parse::Man::DOM::Document;

=head1 Parse::Man::DOM::Document

Represents the document as a whole.

=cut

sub new
{
   my $class = shift;
   return bless { meta => {}, paras => [] }, $class;
}

=head2 $meta = $document->meta( $key )

Returns a C<Parse::Man::DOM::Metadata> object for the named item of metadata.

=over 4

=item * name

The page name given to the C<.TH> directive.

=item * section

The section number given to the C<.TH> directive.

=back

=cut

sub meta
{
   my $self = shift;
   my ( $name ) = @_;
   return $self->{meta}{$name} || die "No meta defined for $name";
}

sub add_meta
{
   my $self = shift;
   my ( $meta ) = @_;
   $self->{meta}{ $meta->name } = $meta;
}

=head2 @paras = $document->paras

Returns a list of C<Parse::Man::DOM::Heading> or C<Parse::Man::DOM::Para> or 
subclass objects, containing the actual page content.

=cut

sub paras
{
   my $self = shift;
   return @{ $self->{paras} };
}

sub append_para
{
   my $self = shift;
   push @{ $self->{paras} }, @_;
}

sub last_para { shift->{paras}[-1] }

sub append_chunk { shift->last_para->append_chunk( @_ ) }

package Parse::Man::DOM::Metadata;

=head1 Parse::Man::DOM::Metadata

Represents a single item of metadata about the page.

=cut

sub new
{
   my $class = shift;
   return bless [ $_[0] => $_[1] ], $class;
}

=head2 $name = $metadata->name

The string name of the metadata

=head2 $value = $metadata->value

The string value of the metadata

=cut

sub name  { shift->[0] }
sub value { shift->[1] }

package Parse::Man::DOM::Heading;
use constant type => "heading";

=head1 Parse::Man::DOM::Heading

Represents the contents of a C<.SH> or C<.SS> heading

=cut

sub new
{
   my $class = shift;
   return bless [ $_[0] => $_[1] ], $class;
}

=head2 $level = $heading->level

The heading level number; 1 for C<.SH>, 2 for C<.SS>

=head2 $text = $heading->text

The plain text string of the heading title

=cut

sub level { shift->[0] }
sub text  { shift->[1] }

package Parse::Man::DOM::Para;

=head1 Parse::Man::DOM::Para

Represents a paragraph of formatted text content. Will be one of the following
subclasses.

=cut

=head2 $filling = $para->filling

Returns true if filling (C<.fi>) is in effect, or false if no-filling (C<.nf>)
is in effect.

=head2 $chunklist = $para->body

Returns a C<Parse::Man::DOM::Chunklist> to represent the actual content of the
paragraph.

=cut

sub filling { shift->{filling} }
sub body    { shift->{body} }

package Parse::Man::DOM::Para::Plain;
use base qw( Parse::Man::DOM::Para );
use constant type => "plain";

=head1 Parse::Man::DOM::Para::Plain

Represent a plain (C<.P> or C<.PP>) paragraph.

=head2 $type = $para->type

Returns C<"plain">.

=cut

sub new
{
   my $class = shift;
   my ( $opts, $body ) = @_;
   return bless { (map { $_ => $opts->{$_} } qw( filling indent )), body => $body }, $class;
}

sub append_chunk { shift->body->append_chunk( @_ ) }

package Parse::Man::DOM::Para::Term;
use base qw( Parse::Man::DOM::Para );
use constant type => "term";

=head1 Parse::Man::DOM::Para::Term

Represents a term paragraph (C<.TP>).

=head2 $type = $para->type

Returns C<"term">.

=cut

sub new
{
   my $class = shift;
   my ( $opts, $term, $definition ) = @_;

   return bless { indent => $opts->{indent}, term => $term, definition => $definition }, $class;
}

=head2 $chunklist = $para->term

Returns a C<Parse::Man::DOM::Chunklist> for the defined term name.

=head2 $chunklist = $para->definition

Returns a C<Parse::Man::DOM::Chunklist> for the defined term definition.

=cut

sub term       { shift->{term} }
sub definition { shift->{definition} }

sub append_chunk
{
   my $self = shift;
   my ( $chunk ) = @_;
   if( !$self->{term_done} and $chunk->isa( "Parse::Man::DOM::Linebreak" ) ) {
      $self->{term_done} = 1;
   }
   elsif( !$self->{term_done} ) {
      $self->term->append_chunk( $chunk );
   }
   else {
      $self->definition->append_chunk( $chunk );
   }
}

package Parse::Man::DOM::Para::Indent;
use base qw( Parse::Man::DOM::Para::Plain );
use constant type => "indent";

=head1 Parse::Man::DOM::Para::Indent

Represents an indented paragraph (C<.IP>).

=head2 $type = $para->type

Returns C<"indent">.

=cut

package Parse::Man::DOM::Chunklist;

=head1 Parse::Man::DOM::Chunklist

Contains a list of C<Parse::Man::DOM::Chunk> objects to represent paragraph
content.

=cut

sub new
{
   my $class = shift;
   return bless { chunks => [ @_ ] }, $class;
}

=head2 @chunks = $chunklist->chunks

Returns a list of C<Parse::Man::DOM::Chunk> objects.

=cut

sub chunks { @{ shift->{chunks} } }

sub append_chunk
{
   my $self = shift;
   push @{ $self->{chunks} }, @_;
}

package Parse::Man::DOM::Chunk;

=head1 Parse::Man::DOM::Chunk

Represents a chunk of text with a particular format applied.

=cut

sub new
{
   my $class = shift;
   return bless [ $_[0] => $_[1] ], $class;
}

sub is_linebreak { 0 }
sub is_space     { 0 }
sub is_break     { 0 }

=head2 $format = $chunk->format

The formatting name in effect for this chunk. One of C<"R">, C<"B">, C<"I"> or
C<"SM">.

=head2 $text = $chunk->text

The plain string value of the text for this chunk.

=cut

sub format { shift->[0] }
sub text   { shift->[1] }

package Parse::Man::DOM::Linebreak;
use base qw( Parse::Man::DOM::Chunk );

sub new
{
   my $class = shift;
   return bless [], $class;
}

sub is_linebreak { 1 }

package Parse::Man::DOM::Space;
use base qw( Parse::Man::DOM::Chunk );

sub new
{
   my $class = shift;
   return bless [], $class;
}

sub is_space { 1 }

package Parse::Man::DOM::Break;
use base qw( Parse::Man::DOM::Chunk );

sub new
{
   my $class = shift;
   return bless [], $class;
}

sub is_break { 1 }

=head1 AUTHOR

Paul Evans <leonerd@leonerd.org.uk>

=cut

0x55AA;
